### ISSUE TITLE
Desktop environment wasn't detected (Hyprland with Wayland)

### ISSUE THREAD

**Issue by FirelightFlagboy** [microsoft/vscode#187436]
Does this issue occur when all extensions are disabled?: Yes

## Versions

- VS Code Version: 

  ```txt
  1.80.0
  660393deaaa6d1996740ff4880f1bad43768c814
  x64
  ```

- OS Version: 

  ```
  OS: Arch Linux x86_64
  Kernel: 6.4.2-arch1-1 
  DE: Hyprland
  ```

- Hyprland version:

  ```
  Hyprland, built from branch main at commit f49af187bc0755f9b5be9376bfd52c0cf12c1e11 dirty (Xdg config home support (2047)).
  Tag: v0.26.0-96-gf49af187

  flags: (if any)

  ```

- Kwallet version:

  ```
  kwallet: 5.107.0-1
  kwalletmanager: 23.04.3-1
  ```

## Description

Prior to updating to `1.80.0` I didn't have that pop-up:

```txt
An OS keyring couldn't be identified for storing the encryption related data in your current desktop environment
```

I'm using `hyprland` a Deskopt Environment (DE) and have also installed `kwallet` & `kwalletmanager` to manage my keyring.

> The `dbus` configuration to use `kwallet` as a keyring work because when a start `vscode` after a boot `kwallet` will ask me to unlock the keyring with my password.

## Steps so far

Following <https://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues> for `linux`.

I start `vscode` like so:

```shell
code --disable-extensions [code-no-ext-verbose-crypt-passwd-kwallet.log](https://github.com/microsoft/vscode/files/12000467/code-no-ext-verbose-crypt-passwd-kwallet.log)
```

In the log, I have the following line indicating the desktop environment is `unknown`:

```log
[36033:0710/095618.511162:VERBOSE1:key_storage_util_linux.cc(54)] Password storage detected desktop environment: (unknown)
```

I've tried to pass the option `password-store`, with the value `kde` as suggested by the documentation:

> If you would rather not use `gnome-keyring`, [...], you could try installing `kwallet` on your system. 
> If you do, you will want to set the `password-store` flag to `kde` to detect the installed `kwallet`.

It result in the following warning message, and the pop-up still present

```shell
$ code --disable-extensions --password-store=kde
Warning: 'password-store' is not in the list of known options, but still passed to Electron/Chromium.
```

I've tried with the value `kwallet`, I still get the warning message (`'password-store' is not in the list of known options ...`) but I no longer have the pop-up

But I can't login to Sync using Github account:

- I've the following notification error:

  ```
  You're running in a KDE environment but the OS keyring is not available for encryption. 
  Ensure you have kwallet running.
  ```

  Strange :thinking: because kwallet is running on my side (remember the pop-up to unlock the keyring) and:

  ```shell
  $ systemctl status | grep kwallet | grep -v grep   
                 │ ├─  684 /usr/bin/kwalletd5 --pam-login 12 14
                     │ └─6240 /usr/bin/kwalletd5
  ```

- The credentials information isn't persisted between restart of `vscode`

## Logs

- [`code --disable-extensions --verbose --vmodule="*/components/os_crypt/*=1" --password-store=kde`](https://github.com/microsoft/vscode/files/12000468/code-no-ext-verbose-crypt-passwd-kde.log)

- [`code --disable-extensions --verbose --vmodule="*/components/os_crypt/*=1" --password-store=kwallet`](https://github.com/microsoft/vscode/files/12000471/code-no-ext-verbose-crypt-passwd-kwallet.log)

- [`code --disable-extensions --verbose --vmodule="*/components/os_crypt/*=1"`](https://github.com/microsoft/vscode/files/12000469/code-no-ext-verbose-crypt-no-passwd.log)

**Comment by TylerLeonhardt** [comment]
So `--password-store=kwallet` appears to be the correct value to tell Electron to look for kwallet.

```
[57938:0710/115101.298898:VERBOSE1:key_storage_linux.cc(122)] Selected backend for OSCrypt: KWALLET
[57938:0710/115101.299344:ERROR:object_proxy.cc(623)] Failed to call method: org.kde.KWallet.isEnabled: object_path= /modules/kwalletd: org.freedesktop.DBus.Error.ServiceUnknown: The name org.kde.kwalletd was not provided by any .service files
[57938:0710/115101.299351:ERROR:kwallet_dbus.cc(100)] Error contacting kwalletd (isEnabled)
[57938:0710/115101.299451:ERROR:object_proxy.cc(623)] Failed to call method: org.kde.KLauncher.start_service_by_desktop_name: object_path= /KLauncher: org.freedesktop.DBus.Error.ServiceUnknown: The name org.kde.klauncher was not provided by any .service files
[57938:0710/115101.299456:ERROR:kwallet_dbus.cc(72)] Error contacting klauncher to start kwalletd
[57938:0710/115101.299459:WARNING:key_storage_linux.cc(208)] OSCrypt tried KWallet but couldn't initialise.
[57938:0710/115101.299551:ERROR:object_proxy.cc(623)] Failed to call method: org.kde.KWallet.close: object_path= /modules/kwalletd: org.freedesktop.DBus.Error.ServiceUnknown: The name org.kde.kwalletd was not provided by any .service files
[57938:0710/115101.299555:ERROR:kwallet_dbus.cc(418)] Error contacting kwalletd (close)
[57938:0710/115101.299613:VERBOSE1:key_storage_linux.cc(142)] OSCrypt did not initialize a backend.
[57938:0710/115101.302863:INFO:CONSOLE(627)] "%cTRACE color: #888 [SecretStorageService] Encryption is not available, falling back to in-memory storage", source: vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/workbench/workbench.desktop.main.js (627)
```

It looks like kwallet isn't registered correctly: `The name org.kde.klauncher was not provided by any .service files` ... I'm not familar with this environment but I wonder if something like this helps:
https://askubuntu.com/questions/617955/problem-with-kde-programs-after-upgrading-to-15-04

**Comment by TylerLeonhardt** [comment]
Can you try `--password-store=kwallet5`

**Comment by FirelightFlagboy** [comment]
Hi @TylerLeonhardt, I've tested with the option `password-store` set to `kwallet5` and I'm not greet with the pop-up :)

**Comment by TylerLeonhardt** [comment]
Great! I will update our doc to say `kwallet5`. Reading the Chromium docs more, that makes a lot of sense.

As for the "automatically detecting" part, if you could please drop your desktop environment configuration in:

https://github.com/microsoft/vscode/issues/187338

We're using that issue to track those.

I have this issue close when our docs are opened.

**Comment by FirelightFlagboy** [comment]
Should I add information to #187338 ?

My concern is hyprland is not tied to `kwallet`, some people use `gnome-keyring`.
What seems interesting is #185212 to auto-detect if `kwallet` or `gnome-keyring` is installed.

**Comment by TylerLeonhardt** [comment]
@FirelightFlagboy yeah https://github.com/microsoft/vscode/issues/187338 is to understand what are common environments to see if there is a configuration that stands out that's worth special casing in Electron. While https://github.com/microsoft/vscode/issues/185212 is for the long tail.

https://github.com/microsoft/vscode/issues/185212 is really interesting esp with kwallet and gnome-keying both implement a [Secret Storage compatible](https://www.freedesktop.org/wiki/Specifications/secret-storage-spec/) layer... so you'll probably get a lot of benefit for that issue.
